### PR TITLE
Fix chained optimizer CUDA graph capture for DDP

### PR DESCRIPTION
**Pull request by wujingyue**
- [x] I, the PR author, have personally reviewed every line of this PR.

## What does this PR do?

Fix optimizer CUDA graph capture for DDP models whose dense and expert optimizers use different gradient-statistics process groups. `ChainedOptimizer` currently converts CUDA gradient norms to Python booleans and passes their aggregate to `math.sqrt`, triggering `operation not permitted when stream is capturing`.

Use tensor arithmetic for both default and separate-group norm aggregation when `optimizer_cuda_graph=True`. Keep `math.sqrt` when CUDA graphs are disabled so eager execution retains its Python-float norm and scalar clipping coefficient precision. Returning a tensor in eager mode changes clipping arithmetic and breaks deterministic training goldens.

Add an optimizer regression test comparing CUDA graph replay with an eager optimizer step across two gradient-statistics process groups. Check that both produce matching norms and parameter updates as gradients change between nonzero and zero values. Also verify that eager default and MTP group norms exactly match the previous scalar aggregation for nonzero, partially zero, and all-zero gradients.

## Validation

- Reproduced the capture failure against unpatched code.
- The capture/replay regression passes on each of four GB200 GPUs.
- The eager regression fails on the previous PR head for both default and MTP norm groups on all eight H100 ranks.
- All eight focused chained-optimizer tests pass on each of eight H100 GPUs, including CUDA graph capture/replay.
- Both previously failing H100 recipes pass exact loss and zero-gradient-count comparisons against unchanged goldens: `gpt3_mcore_te_tp2_pp1_te_a2a_ovlp_8experts_etp1_ep4` and `gpt3_moe_mcore_te_ep8_resume_torch_dist_muon`. Muon checkpoint-resume validation also passes.
- The GB200 `_1node` Muon recipe awaits the CI rerun.
- Black, isort, Pylint, and Ruff checks pass.
- Use `determinism-exempt` for the file-level determinism gate: this changes norm aggregation for graph capture; the multi-tensor and fused optimizer kernels are unchanged. The regression belongs with optimizer tests.

```bash
python -m torch.distributed.run --standalone --nproc-per-node=4 -m pytest -q \
  tests/unit_tests/optimizer/test_chained_optimizer.py
```

### CODE REVIEW

**Review by guihong-nv — APPROVED** [review]

File: tests/unit_tests/optimizer/test_chained_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                param.data.zero_()
+                param.grad.fill_(value)
+            _, expected_norm, _ = optimizer.step()
+            expected_params = [param.detach().clone() for param in params]
+
+            for param, value in zip(params, values):
+                param.data.zero_()
+                param.grad.fill_(value)
+            graph.replay()
+            torch.testing.assert_close(grad_norm, expected_norm)
```
Comment (skyw) [inline tests/unit_tests/optimizer/test_chained_optimizer.py:100]: torch.testing.assert_close default is not very close. Should graph replay match eager exactly? if yes, recommend to add atol=0,rtol=0.

Comment (wujingyue) [reply tests/unit_tests/optimizer/test_chained_optimizer.py:100]: > Should graph replay match eager exactly?

Not in general. IIRC, TE's FusedAdam uses a different kernel for graph mode. I've yet to try to understand the reason.

Comment (skyw) [reply tests/unit_tests/optimizer/test_chained_optimizer.py:100]: I see.
It could be similar reason as pytorch version, see https://github.com/pytorch/pytorch/blob/6325c81b68725776d0bc5d6854e6bb16591c2165/torch/optim/adam.py#L478? usually mutable variable like step count are on CPU
